forge-game/src/main/java/forge/game/phase/PhaseHandler.java

```diff
@@ -1024,132 +1019,147 @@ public void startFirstTurn(Player goesFirst, Runnable startGameHook) {
             startGameHook.run();
             givePriorityToPlayer = true;
         }
+    }
 
-        // MAIN GAME LOOP
-        while (!game.isGameOver()) {
-            if (givePriorityToPlayer) {
-                if (DEBUG_PHASES) {
-                    sw.start();
-                }
-
-                game.fireEvent(new GameEventPlayerPriority(playerTurn, phase, getPriorityPlayer()));
-                List<SpellAbility> chosenSa = null;
-
-                int loopCount = 0;
-                do {
-                    if (checkStateBasedEffects()) {
-                        // state-based effects check could lead to game over
-                        return;
-                    }
-                    game.stashGameState();
-
-                    chosenSa = pPlayerPriority.getController().chooseSpellAbilityToPlay();
+    public void startFirstTurn(Player goesFirst) {
+        startFirstTurn(goesFirst, null);
+    }
+    public void startFirstTurn(Player goesFirst, Runnable startGameHook) {
+        setupFirstTurn(goesFirst, startGameHook);
+        mainGameLoop();
+    }
 
-                    // this needs to come after chosenSa so it sees you conceding on own turn
-                    if (playerTurn.hasLost() && pPlayerPriority.equals(playerTurn) && pFirstPriority.equals(playerTurn)) {
-                        // If the active player has lost, and they have priority, set the next player to have priority
-                        System.out.println("Active player is no longer in the game...");
-                        pPlayerPriority = game.getNextPlayerAfter(getPriorityPlayer());
-                        pFirstPriority = pPlayerPriority;
-                    }
+    public void mainGameLoop() {
+        // MAIN GAME LOOP
+        while (!game.isGameOver() && !(game.getAge() == GameStage.RestartedByKarn)) {
+            mainLoopStep();
+        }
+    }
 
-                    if (chosenSa == null) {
-                        break; // that means 'I pass'
-                    }
-                    if (DEBUG_PHASES) {
-                        System.out.print("... " + pPlayerPriority + " plays " + chosenSa);
-                    }
+    public void mainLoopStep() {
+        StopWatch sw = new StopWatch();
```

Comment:
It doesn't make sense to create a new object each step
maybe turn it into a class field?